pkgs/top-level/unit-files.nix

```diff
@@ -0,0 +1,105 @@
+    else
+      mapAttrs
+        (name: _: baseDirectory + "/${shard}/${name}/package.nix")
+        (readDir (baseDirectory + "/${shard}"));
+
+  # Merge the name mappings of all shards between the indices start (inclusive) and end (exclusive)
+  # This ignores any potential duplicates for efficiency
+  # Invariant: start < end
+  # Type: Int -> Int -> AttrsOf Path
+  binaryMerge = start: end:
+    if end - start >= 2 then
+      # If there's more than 2 elements, split the range in two, recurse on each part and merge the result
+      # The binaryMerge invariant is satisfied because each half will have at least 1 element
+      binaryMerge start (start + (end - start) / 2)
+      // binaryMerge (start + (end - start) / 2) end
+    else
+      # Otherwise there will be exactly 1 element due to the invariant, in which case we just return it directly
+      namesForShard (elemAt shards start);
+
+  # We could get the same result with a simple
+  #
+  #   builtins.foldl' (acc: entry: acc // namesForShard entry) {} shards;
+  #
+  # but this has rather poor evaluation performance for a large number of shards.
+  # This is because `l // r` has complexity O(l + r) with l and r being the number attributes on the left and right side respectively.
+  # So e.g. if we have n shards each with m entries, the above fold would be evaluated like this:
+  #     /\     O((n-1)m + m) = O(nm)
+  #    ⋰  m    ... (n times)
+  #   /\       O(2m + m)     = O(3m)
+  #  /\ m      O(m + m)      = O(2m)
+  # m  m
+  #
+  # Giving a total complexity of O(n²m)
+  #
+  # To improve this we make use of the fact that `//` is associative, allowing us to rearrange evaluation as a binary tree:
+  #         / \         O(1 * (0.5nm + 0.5nm)) = O(nm)
+  #      ⋰       ⋱      ... (log2(n) times)
+  #    / \       / \    O(n / 4 * (2m + 2m))   = O(nm)
+  #  /\   /\   /\   /\  O(n / 2 * (m + m))     = O(nm)
+  # m  m m  m m  m m  m
+  #
+  # Giving a total complexity of only O(log2(n)nm)
+
+  # Another potential approach is for `namesForShard` to return [ { name, value } ] instead, and then use builtins to merge it together:
+  #
+  #   listToAttrs (concatMap namesForShard shards)
+  #
+  # This way the shard merging can be done by Nix's C++ code.
+
+  # We benchmark the three approaches with the following command with 300 runs:
+  #
+  #   NIX_SHOW_STATS=1 nix-instantiate --eval --strict 'builtins.seq (import pkgs/top-level/unit-files.nix) null'
+  # Approach     | cpuTime [ms] | gc.totalBytes [MB]
+  # linear merge | 33.4 ± 4.8   | 126.075
+  # concatMap    | 21.9 ± 2.6   |   5.824
+  # binary merge | 16.6 ± 2.5   |   7.102
+  #
+  # The lowest evaluation time is achieved with the binary merge algorithm, even despite the slightly higher memory usage.
+  # However if https://github.com/NixOS/nix/issues/7676 were implemented in the future, the concatMap-based approach could be much more performant.
+  # Alternatively, other Nix builtins to speed up this computation would help too.
```

Comment:
@piegamesde Turns out it's a bit slower

Reply:
Unfortunate. We'd need a builtin like `concatAttrs` and maybe also `concatMapAttrs` for this then.

What about exposing your `binaryMerge` as `lib.concatAttrs`, so that a) people don't have to ad-hoc define it like done here and b) there is a greater chance of elevating this into a builtin later on?

Reply:
The interface that would have to be exposed to get the same efficiency is `concatAttrs :: [String] -> (String -> AttrsOf a) -> AttrsOf a`, but the arguably expected interface would instead be `concatAttrs :: [AttrsOf a] -> AttrsOf a`. Because of this I'd prefer leaving it out of the discussion for now.

Reply:
`concatMapAttrs` being a `foldl'`, it is very close to the _linear merge_ implementation, so I would not expect it to perform better as implemented. Turning it into an optimized primitive in nix itself may be worth it though.

> I'm unhappy about that name [`concatMapAttrs`]

I am to blame for that, see discussion in
 - https://github.com/NixOS/nixpkgs/pull/201527

I'd be open to discuss it there if you can think of a better name that follows a useful naming convention. It's less than a year old, so I think it's worth renaming _if_ it produces a significant improvement.

Reply:
Fwiw Nix, as implemented in `NixOS/nix`, can't make use of the property that the attributes are already in lexicographical order. The internal ordering of attributes is an arbitrary ordering on internalized strings. Nonetheless, I expect a native implementation to be able to beat what we can do in the language.

Reply:
Oh, I didn't know `concatMapAttrs` was actually a thing in `lib`, the one I was describing was some made-up thing with a made-up name. The difference between the two ist, that the one in nixpkgs takes an attribute set as input while what we currently need is a list as input.

Reply:
> Interesting, so `concatAttrs (map namesForShard shards)` would have worse performance then?

It does additionally have to allocate a ~700 item list, though this probably doesn't have a significant performance decrease.